File: android/app/src/main/java/com/adasa/foodplan/domain/model/Ingredient.kt
package com.adasa.foodplan.domain.model

enum class IngredientCategory(val displayName: String, val emoji: String) {
    FRUIT_VEG(      "Fruit & Vegetables",   "🥦"),
    MEAT(           "Meat",                 "🥩"),
    FISH(           "Fish & Seafood",       "🐟"),
    DAIRY_EGGS(     "Dairy & Eggs",         "🥚"),
    CHEESE(         "Cheese",               "🧀"),
    GRAINS(         "Grains & Rice",        "🌾"),
    BREAD_BAKERY(   "Bread & Bakery",       "🍞"),
    DRY_GOODS(      "Dry Goods & Pasta",    "🫙"),
    NUTS(           "Nuts & Seeds",         "🥜"),
    CANNED(         "Canned & Preserved",   "🥫"),
    FROZEN(         "Frozen",               "🧊"),
    OILS_SAUCES(    "Oils & Sauces",        "🫗"),
    SPICES(         "Spices & Herbs",       "🌿"),
    DRINKS(         "Drinks",               "🥤"),
    SUPPLEMENT(     "Supplements",          "💊"),
    OTHER(          "Other",                "📦"),
}

data class Ingredient(
    val id: String = "",
    val name: String = "",
    val category: IngredientCategory = IngredientCategory.OTHER,
    val kcalPer100g: Double = 0.0,
    val proteinPer100g: Double = 0.0,
    val fatPer100g: Double = 0.0,
    val carbsPer100g: Double = 0.0,
    val source: IngredientSource = IngredientSource.LABEL,
    val steps: List<String> = emptyList(),
)

enum class IngredientSource(val displayName: String) {
    LABEL("Produktetikett"),
    LIVSMEDELSVERKET("Livsmedelsverket"),
    CALCULATED("Beräknat"),
    BARCODE("Streckkod")
}

File: flutter-app/lib/domain/model/ingredient.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'ingredient.freezed.dart';

// ── Enums ─────────────────────────────────────────────────────────────────────

enum IngredientCategory {
  fruitVeg('Fruit & Vegetables', '🥦', 'FRUIT_VEG'),
  meat('Meat', '🥩', 'MEAT'),
  fish('Fish & Seafood', '🐟', 'FISH'),
  dairyEggs('Dairy & Eggs', '🥚', 'DAIRY_EGGS'),
  cheese('Cheese', '🧀', 'CHEESE'),
  grains('Grains & Rice', '🌾', 'GRAINS'),
  breadBakery('Bread & Bakery', '🍞', 'BREAD_BAKERY'),
  dryGoods('Dry Goods & Pasta', '🫙', 'DRY_GOODS'),
  nuts('Nuts & Seeds', '🥜', 'NUTS'),
  canned('Canned & Preserved', '🥫', 'CANNED'),
  frozen('Frozen', '🧊', 'FROZEN'),
  oilsSauces('Oils & Sauces', '🫗', 'OILS_SAUCES'),
  spices('Spices & Herbs', '🌿', 'SPICES'),
  drinks('Drinks', '🥤', 'DRINKS'),
  supplement('Supplements', '💊', 'SUPPLEMENT'),
  other('Other', '📦', 'OTHER');

  const IngredientCategory(this.displayName, this.emoji, this._firestoreName);

  final String displayName;
  final String emoji;
  final String _firestoreName;

  String get firestoreName => _firestoreName;

  static IngredientCategory fromFirestore(String? s) {
    if (s == null) return IngredientCategory.other;
    try {
      return IngredientCategory.values.firstWhere(
        (e) => e._firestoreName == s.toUpperCase(),
      );
    } catch (_) {
      return IngredientCategory.other;
    }
  }
}

enum IngredientSource {
  label,
  livsmedelsverket,
  calculated,
  barcode;

  String get firestoreName => name.toUpperCase();

  static IngredientSource fromFirestore(String? s) {
    if (s == null) return IngredientSource.label;
    try {
      return IngredientSource.values.firstWhere(
        (e) => e.firestoreName == s.toUpperCase(),
      );
    } catch (_) {
      return IngredientSource.label;
    }
  }
}

// ── Domain model ──────────────────────────────────────────────────────────────

@freezed
class Ingredient with _$Ingredient {
  const factory Ingredient({
    required String id,
    required String name,
    @Default(IngredientCategory.other) IngredientCategory category,
    @Default(0.0) double kcalPer100g,
    @Default(0.0) double proteinPer100g,
    @Default(0.0) double fatPer100g,
    @Default(0.0) double carbsPer100g,
    @Default(IngredientSource.label) IngredientSource source,
    @Default(<String>[]) List<String> steps,
  }) = _Ingredient;
}
